nomad/drainer/watch_nodes.go

```diff
@@ -241,15 +245,6 @@ func (w *nodeDrainWatcher) getNodesImpl(ws memdb.WatchSet, state *state.StateSto
 
 		node := raw.(*structs.Node)
 		resp[node.ID] = node
-		if maxIndex < node.ModifyIndex {
-			maxIndex = node.ModifyIndex
```

Comment:
This was unreachable right?! It seems like if a `Node.ModifyIndex > Index("nodes")` we have a bug?! I know I :heavy_check_mark: the original PR, but I can't reason about what this did...

Reply:
> Since this wakes up any time any node is updated (both before and after this patch AFAICT), I wonder if we should rate limit it a bit to avoid iterating over every node as fast as possible. 

There's a rate limiter set to 100 state queries / second, which is shared between the job watcher and node watcher. The job watcher does a blocking query on allocs for just the tracked jobs.

> Also I don't even know if using an index at all is useful here if we're just waking up on every "nodes" table change. I think we could just block on the WatchSet ourselves. Performance-wise I don't expect this to really matter though, especially if we start coalescing iterations over a short timeframe.

From a code structure standpoint `(*StateStore).BlockingQuery` gives us nice QoL things like exiting when the context is canceled that we'd have to implement separately for a long-lived poll like this. I did some very rough testing and even running a very tight test with a single node I'm seeing queries about ~100ms apart. The real limiting factor is going to be how quickly we can write raft entries and the rest of the time it'll just block. So in practice I don't think we're going to keep this particular goroutine spinning on-CPU all that much.

That being said, I'm going to be taking a deep dive into the job watcher as my next task (to try and finally land https://github.com/hashicorp/nomad/pull/14348), so I can do some more detailed performance analysis as part of that work.